lightning/src/routing/onion_message.rs

```diff
@@ -0,0 +1,280 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! [Onion message] pathfinding lives here.
+//!
+//! Finding paths for onion messages is necessary for retrieving invoices and fulfilling invoice
+//! requests from [offers]. It differs from payment pathfinding in that channel liquidity, fees, and
+//! knobs such as `htlc_maximum_msat` do not factor into path selection -- onion messages require a
+//! peer connection and nothing more. However, we still use the network graph because onion messages
+//! between channel peers are likely to be prioritized over those between non-channel peers.
+//!
+//! [Onion message]: crate::onion_message
+//! [offers]: <https://github.com/lightning/bolts/pull/798>
+
+use bitcoin::secp256k1::{self, PublicKey};
+
+use ln::channelmanager::ChannelDetails;
+use routing::gossip::{NetworkGraph, NodeId};
+use util::logger::Logger;
+
+use alloc::collections::BinaryHeap;
+use core::cmp;
+use core::fmt;
+use core::ops::Deref;
+use prelude::*;
+
+/// Finds a route from us to the given `destination` node.
+/// If we have private channels, it may be useful to fill in `first_hops` with the results from
+/// [`ChannelManager::list_usable_channels`]. If `first_hops` is not filled in, connected peers
+/// without public channels will not be forwarded over.
+///
+/// [`ChannelManager::list_usable_channels`]: crate::ln::channelmanager::ChannelManager::list_usable_channels
+pub fn find_path<L: Deref, GL: Deref>(
+	our_node_pubkey: &PublicKey, destination: &PublicKey, network_graph: &NetworkGraph<GL>, first_hops: Option<&[&ChannelDetails]>, logger: L
+) -> Result<Vec<PublicKey>, Error> where L::Target: Logger, GL::Target: Logger
+{
+	log_trace!(logger, "Searching for an onion message path from origin {} to destination {} and {} first hops {}overriding the network graph",
+		our_node_pubkey, destination, first_hops.map(|hops| hops.len()).unwrap_or(0),
+		if first_hops.is_some() { "" } else { "not " });
+	let graph_lock = network_graph.read_only();
+	let network_channels = graph_lock.channels();
+	let network_nodes = graph_lock.nodes();
+	let our_node_id = NodeId::from_pubkey(our_node_pubkey);
+	let dest_node_id = NodeId::from_pubkey(destination);
+	if our_node_id == dest_node_id { return Err(Error::InvalidDestination) }
+
+	// Add our start and first-hops to `frontier`.
+	let start = NodeId::from_pubkey(&our_node_pubkey);
+	let mut valid_first_hops = HashSet::new();
+	let mut frontier = BinaryHeap::new();
+	frontier.push(PathBuildingHop { cost: 0, node_id: start, parent_node_id: start });
```

Comment:
Can define the reasoning behind `PathBuildingHop::cost`. I don't know if the onion bandwidth we're expecting from hops on the network is scaled on the channel capacity. The most rule of thumb metric I would guess.